crfs.go

```diff
@@ -211,20 +782,85 @@ func (n *node) Readlink(ctx context.Context, req *fuse.ReadlinkRequest) (string,
 	return n.te.LinkName, nil
 }
 
-// Read reads data from a regular file n.
-//
-// See https://godoc.org/bazil.org/fuse/fs#HandleReader
-func (n *node) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadResponse) error {
-	sr, err := n.fs.r.OpenFile(n.te.Name)
-	if err != nil {
-		return err
+func (n *node) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fspkg.Handle, error) {
+	h := &nodeHandle{
+		n:     n,
+		isDir: req.Dir,
 	}
+	resp.Handle = h.HandleID()
+	if !req.Dir {
+		var err error
+		h.sr, err = n.sr.OpenFile(n.te.Name)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return h, nil
+}
+
+// nodeHandle is a node that's been opened (opendir or for read).
+type nodeHandle struct {
+	n     *node
+	isDir bool
+	sr    *io.SectionReader // of file bytes
+
+	mu            sync.Mutex
+	lastChunkOff  int64
+	lastChunkSize int
+	lastChunk     []byte
+}
+
+func (h *nodeHandle) HandleID() fuse.HandleID {
+	return fuse.HandleID(uintptr(unsafe.Pointer(h)))
+}
+
+func (h *nodeHandle) chunkData(offset int64, size int) ([]byte, error) {
+	h.mu.Lock()
+	if h.lastChunkOff == offset && h.lastChunkSize == size {
+		defer h.mu.Unlock()
+		//log.Printf("cache HIT, chunk off=%d/size=%d", offset, size)
+		return h.lastChunk, nil
+	}
+	h.mu.Unlock()
+
+	log.Printf("reading chunk for offset=%d, size=%d", offset, size)
+	buf := make([]byte, size)
+	n, err := h.sr.ReadAt(buf, offset)
+	log.Printf("... ReadAt = %v, %v", n, err)
+	if err == nil {
+		h.mu.Lock()
+		h.lastChunkOff = offset
+		h.lastChunkSize = size
+		h.lastChunk = buf
+		h.mu.Unlock()
+	}
+	return buf, err
+}
+
+// See https://godoc.org/bazil.org/fuse/fs#HandleReader
+func (h *nodeHandle) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadResponse) error {
+	n := h.n
 
 	resp.Data = make([]byte, req.Size)
-	nr, err := sr.ReadAt(resp.Data, req.Offset)
-	if nr < req.Size {
-		resp.Data = resp.Data[:nr]
+	nr := 0
+	offset := req.Offset
+	for nr < req.Size {
+		ce, ok := n.sr.ChunkEntryForOffset(n.te.Name, offset+int64(nr))
+		if !ok {
+			break
+		}
+		if debug {
```

Comment:
I don't see `debug` being set anywhere, so I'm guessing you just twiddle this manually in the source? Is it still used?

Reply:
Yeah, I twiddle it while debugging. And because it's a const, it compiles out otherwise.